src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -144,17 +147,45 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
             }
         }
 
+        public bool TryUpdateOutputWindow(int bytes)
+        {
+            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+        }
+
         public override void Abort(ConnectionAbortedException abortReason)
         {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
             base.Abort(abortReason);
 
             // Unblock the request body.
             RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
         }
 
-        public bool TryUpdateOutputWindow(int bytes)
+        protected override void ApplicationAbort()
         {
-            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+            Log.ApplicationAbortedConnection(ConnectionId, TraceIdentifier);
+            var abortReason = new ConnectionAbortedException(CoreStrings.ConnectionAbortedByApplication);
+            ResetAndAbort(abortReason, Http2ErrorCode.CANCEL);
+        }
+
+        private void ResetAndAbort(ConnectionAbortedException abortReason, Http2ErrorCode error)
+        {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
+            // If this was an app initiated Abort then we need to inform the client.
+            _http2Output.WriteRstStreamAsync(endReceived: EndStreamReceived, error).GetAwaiter().GetResult();
```

Comment:
Having `HttpContext.Abort()` block on I/O for any reason is new. I don't think we should do it. Fire and forget is better. This write can't fail since no cancellation token is passed in anyway.